[stable/sonatype-nexus] a significant refactor of the nexus chart that allows for easy backups and proxy

jeff-knurek opened the pull request:
Makes the chart more production ready from install

Commit:
[stable/sonatype-nexus] a significant refactor of the nexus chart that allows for easy backups and proxy

Makes the chart more production ready from install

jeff-knurek commented:
/assign @rjkernick

jeff-knurek commented:
/assign @sameersbn

rjkernick commented:
Looks good. Thanks for the refactor!

unguiculus commented on stable/sonatype-nexus/templates/_helpers.tpl:
```diff
@@ -2,15 +2,33 @@
 We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
 */}}
-{{- define "fullname" -}}
+{{- define "nexus.fullname" -}}
```
While you're at it: Please update templates to those create by `helm create` with Helm 2.8+.

unguiculus commented on stable/sonatype-nexus/templates/statefulset.yaml:
```diff
@@ -0,0 +1,140 @@
+apiVersion: apps/v1beta1
```
`apps/v1beta2`

unguiculus replied:
Can you explain why you switched to statefulset? It doesn't seem obvious to me.

unguiculus commented on stable/sonatype-nexus/README.md:
```diff
@@ -1,14 +1,35 @@
 ## Prerequisites
 
-- Kubernetes 1.6+
+- Kubernetes 1.7+ with Beta APIs enabled
```
1.8+

We officially support the current and previous minor version of Kubernetes.

unguiculus commented on stable/sonatype-nexus/templates/pvc.yaml:
```diff
@@ -2,23 +2,14 @@
-{{- if (eq "-" .Values.persistence.storageClass) }}
-  storageClassName: ""
-{{- else }}
+      storage: {{ .Values.persistence.storageSize }}
```
Pleasse keep the existing storage class configuration. This is what we recommend as best practice.

See #1869

unguiculus commented on stable/sonatype-nexus/templates/proxy-svc.yaml:
```diff
@@ -0,0 +1,15 @@
+      protocol: TCP
+      targetPort: {{ .Values.nexus.externalPort }}
+  selector:
+    app: {{ template "nexus.fullname" . }}
```
```yaml
app: {{ template "nexus.name" . }}
release: {{ .Release.Name }}
```

unguiculus commented on stable/sonatype-nexus/values.yaml:
```diff
@@ -1,68 +1,82 @@
-  # requests:
-  #  cpu: "100m"
-  #  memory: "128Mi"
+  storageClass: ""
```
Keep the existing comment on storage class configuration.

Commit:
[sonatype-nexus] add in the template changes from Helm 2.8.1

Commit:
[sonatype-nexus] implement as a deployment instead of statefulset

stable storage is a requirement, but couldn't find a reason why it still needed to be a statefulset

Commit:
[sonatype-nexus] increase k8s version as prerequisite

Commit:
[sonatype-nexus] fix the other PR comments, namely keeping the storage class settings consistent

jeff-knurek commented:
@unguiculus thanks for the helpful feedback on the review

unguiculus commented on stable/sonatype-nexus/requirements.yaml:
```diff
@@ -0,0 +1,5 @@
+dependencies:
```
I don't think we should add cert-manager as a dependency. Imagine we'd do this with any chart that could expose a service. This should be installed separately and usually not on a per-app basis.

unguiculus commented on stable/sonatype-nexus/templates/proxy-svc.yaml:
```diff
@@ -0,0 +1,16 @@
+    - name: {{ .Values.nexusProxy.name }}
+      port: {{ .Values.nexus.externalPort }}
+      protocol: TCP
+      targetPort: {{ .Values.nexus.externalPort }}
```
Should certainly not be the external port. I'd suggest you hard-code container ports and reference them by name. Changing them would break things unless you reconfigure Nexus itself with the changed ports.

jeff-knurek replied:
I think there was some poor naming of the ports (fixed in latest commit). But I think a great benefit of such options in Helm is that it's easy to use a different image that uses a different port. Hopefully the setup is a little more clear on usage.

unguiculus commented on stable/sonatype-nexus/values.yaml:
```diff
@@ -1,68 +1,85 @@
-  type: LoadBalancer
-  externalPort: 8081
+nexus:
+  imageName: quay.io/travelaudience/docker-nexus
```
You changed the image from some random image to another random image. Why don't we use the official image?

https://hub.docker.com/r/sonatype/nexus3/

jeff-knurek replied:
The official image makes it difficult to provide restore capabilities, and the Nexus service needs to be stopped in order for them to be applied. 

This is better documented here: https://github.com/travelaudience/kubernetes-nexus#restore _(this documentation is linked to at the bottom of the README)_

Commit:
[sonatype-nexus] rename the port values, and make it clearer how they are used

Commit:
[sonatype-nexus] remove dependency of cert-manager

bbriggs commented:
Hey @jeff-knurek, we at @cotalabs are very interested in the changes you are proposing for this PR but want to know if you're still working on it. If there's anything we can help with, please let me know. I'd love to assist in any work to get this merged.

Commit:
[sonatype-nexus] during merge conflicts, almost lost latest change of pv annotations

jeff-knurek commented:
well, I've applied the changes originally requested, _(or commented on why they weren't changed)_, and I just now updated the PR with the latest changes from upstream. So I guess it's just waiting approval and/or feedback....

* there has been some feedback against using yet another custom docker image, and I definitely agree on that topic. But until the official nexus image allows for backup/restore procedure, I find it hard to agree that it's better. And getting that functionality into the official image isn't something I plan on doing anytime soon.

apenney commented:
I don't know if it'll help, but I'll start a conversation with our sonatype contacts to see if they can help us get that stuff into the official container.  It won't be fast, I'm sure, but I'll see!

Commit:
[sonatype-nexus] update nexus version to 3.9.0

whyjustin commented:
> I don't know if it'll help, but I'll start a conversation with our sonatype contacts to see if they can help us get that stuff into the official container. It won't be fast, I'm sure, but I'll see!

@jeff-knurek @apenney Hi, I'm a Sonatype contact - product owner who has charge of the Docker images for Nexus Repository Manager. I will look into the comment threads to see what I can grok but if there are specific changes that would help you out please email me at jyoung at sonatype.

bbriggs commented:
@jeff-knurek Maybe I'm missing it, but is there a way to disable deploying the backup container/service? If I want to test this locally but skip using the backup container for those tests, I don't see a boolean to let me do that in the current setup.

jeff-knurek commented:
@bbriggs that hadn't been considered an option, I'll look into that today

@whyjustin I'll send you an email with some more details

bbriggs commented:
Sounds great. I'm doing some work to get the chart running in Minikube. If I finish before this gets merged, would you mind me sending you the changes to consider?

Commit:
[sonatype-nexus] add option for securityContext to allow for official sonatype/nexus3 image to start up

Commit:
[sonatype-nexus] add option for disabling the backup container (and it's resources)

bbriggs commented:
@jeff-knurek When setting the `nexusBackup.enabled` to false, my helm deployment failed to start due to a bad mount. It looked like `nexus` was trying to improperly mount the volume `nexus-backup` for some reason. Here's the change I made to get it working on my end:

```
diff --git a/stable/sonatype-nexus/templates/deployment.yaml b/stable/sonatype-nexus/templates/deployment.yaml
index f4a25830..677fe772 100644
--- a/stable/sonatype-nexus/templates/deployment.yaml
+++ b/stable/sonatype-nexus/templates/deployment.yaml
@@ -37,8 +37,10 @@ spec:
           volumeMounts:
             - mountPath: /nexus-data
               name: nexus-data
+            {{- if .Values.nexusBackup.enabled }}
             - mountPath: /nexus-data/backup
               name: nexus-backup
+            {{- end }}
         - name: nexus-proxy
           image: {{ .Values.nexusProxy.imageName }}:{{ .Values.nexusProxy.imageTag }}
           imagePullPolicy: {{ .Values.nexusProxy.imagePullPolicy }}
@@ -140,6 +142,7 @@ spec:
           {{- else }}
           emptyDir: {}
           {{- end }}
+        {{- if .Values.nexusBackup.enabled }}
         - name: nexus-backup
           {{- if .Values.nexusBackup.persistence.enabled }}
           persistentVolumeClaim:
@@ -147,3 +150,4 @@ spec:
           {{- else }}
           emptyDir: {}
           {{- end }}
+        {{- end }}
```

bbriggs commented:
Hello, any progress on this?

jeff-knurek commented:
sorry, have been on holiday for the past week - if I don't get it added today, i will tomorrow

Commit:
[sonatype-nexus] add option for using an existing PV

Commit:
[sonatype-nexus] use an empty dir for backup if backup container isn't enabled

Commit:
[sonatype-nexus] update NOTES.txt to use the actual value of admin pwd instead of hardcoded default

unguiculus commented:
ping @rjkernick

rjkernick commented on stable/sonatype-nexus/templates/deployment.yaml:
```diff
@@ -1,62 +1,149 @@
-            {{- if .Values.docker.enabled }}
-            - containerPort: {{ .Values.docker.port }}
-            {{- end }}
-          livenessProbe:
```
Is there a particular reason the liveness/readiness probes were removed? If not, can you please add them back in?

Commit:
[sonatype-nexus] add the liveness/readiness probe back in

rjkernick commented:
/ok-to-test

Commit:
[sonatype-nexus] Backup is currently only works with GCP at the moment, so disable it by default

Commit:
[sonatype-nexus] Add a link to the repo that is used for the nexusBackup container

jeff-knurek commented:
/retest

unguiculus commented:
/lgtm

rjkernick commented:
/lgtm

Nodraak commented:
Hi!

This PR broke the ingress configuration.

**The ingress' key `ingress.spec.rules.http.paths.path` should be `/` instead of `/*`.**

~**Should I submit a PR ?**~ Edit: Submitted as #5205

(Ref: https://github.com/kubernetes/ingress-nginx/issues/250#issuecomment-304929353)

*(Note: I am using Kubernetes v1.10.0 and nginx-ingress quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.11.0)*

jeff-knurek commented:
yeah, you should submit another PR at this point, but that change makes sense

wjeanneau commented:
Hi, I'm not completely sure about how to use nexus-proxy, but do you think it could be good to have an option to disable it ?
